libs/@local/advanced-types/src/distribute.ts

```diff
@@ -17,3 +17,48 @@ export type DistributiveField<T, K> = T extends unknown
     ? T[K]
     : never
   : never;
+
+/**
+ * Replace properties in a union type, distributing the replacement across each member of the union,
+ * and enforcing mutual exclusion with `?: never` properties.
+ *
+ * This is equivalent to `ExclusiveUnion<DistributiveOmit<T, keyof R> & R>` but more concise.
+ *
+ * @example
+ * ```typescript
+ * type Original = { a: string; old: number } | { b: string; old: number };
+ * type Replaced = DistributiveReplaceProperties<Original, { new: boolean }>;
+ * // Result: { a: string; new: boolean; b?: never } | { b: string; new: boolean; a?: never }
+ * ```
+ */
+export type DistributiveReplaceProperties<T, R> = T extends unknown
+  ? Omit<T, keyof R> & R
+  : never;
+
+/**
+ * Enforces mutual exclusion on a union type by adding `?: never` properties for discriminating keys
+ * that don't exist in each variant.
+ *
+ * Given a union like `{ a: string } | { b: number }`, this will produce:
+ * `{ a: string; b?: never } | { a?: never; b: number }`
+ *
+ * This makes TypeScript enforce that you can't pass `{ a: "x", b: 1 }` to a function expecting this union.
+ * Without this, TypeScript allows excess properties when assigning to variables (not inline literals).
+ *
+ * @param T - The union type to make exclusive
+ * @param Keys - Optional explicit list of discriminator keys. If not provided, uses all keys from all union members.
+ *
+ * @example
+ * ```typescript
+ * type Params = { query: string } | { filter: Filter };
+ * type ExclusiveParams = ExclusiveUnion<Params>;
+ * // Result: { query: string; filter?: never } | { query?: never; filter: Filter }
+ *
+ * const valid: ExclusiveParams = { query: "test" }; // ✓
+ * const invalid: ExclusiveParams = { query: "test", filter: {...} }; // ✗ Type error
+ * ```
+ */
+export type ExclusiveUnion<
```

Comment:
This also works, the "idiomatic" way of solving this in JS AFAIR is using a tagged field (e.g. `_tag`), which makes it so that you don't need to use a type utility and is therefore computationally less expensive.